Piece, movement patterns and move worker

ValleSten opened the pull request:
This Pull Request contains:

- The class Piece that represents a piece that can have a customized movement pattern
- The class MoveWorker that handles the movement of the pieces on the board
- Some changes and extensions to the class Chessboard
- Several additional tests in ChessBoardTests.cs.

The movement patterns are not easy to understand how they should be implemented, we talked a bit about this during this weeks supervision meeting, but should probably discuss it a bit more thoroughly. The same applies for how all the valid moves are retrieved: it's not easy to understand, but we are happy to talk about it.

Everything works as intended (read: the tests pass), but there are many optimizations that could be done with the implementations included in this PR. However, since this is supposed to be merged with the separate work done by @sjohan99, @jonathanfransson99 and @Joakim-Holmberg we decided to leave these alterations until we know more about the full scope of how the entities using these classes work. In other words, we are aware that it will require some refactoring :)

Happy reviewing!

sjohan99 reviewed (commented):
Good work! This is a complicated part of the code so it's only natural that the code also gets a bit complicated. I think it's possible to simplify it without too much work though. I've added some specific comments on what I think could be done. One thing that I mention multiple times is making "early exits" (if the term is unfamiliar, check the two first [code blocks](https://medium.com/swlh/return-early-pattern-3d18a41bba8)). I think that can help a lot, as well as some functional decomposition in a some cases.

ValleSten commented:
Thanks for your insight! I think I have resolved all the issues you mentioned, at least the code looks a lot cleaner now ;). I've implemented early exits in the places you mentioned (and some additional places I found, great tip!) as well as some refactoring of other methods and fields. I've also added documentation in places where it was missing.

Regarding your comment on "Test_Custom_Piece": small tests are always easier to understand, but with such a complicated movement pattern as that piece has, I think using an editor to visualize the movements would be of ease even for smaller tests :sweat_smile:

sjohan99 reviewed (approved):
Nice! I think that looks a lot better!
